### PR TITLE
Workflows Base Module

### LINKED ISSUE #237: accept dihedrals_kws dictionary as input for use with workflows base and registry
link PR #229 

* after merging PR #217 
* modify workflows.dihedral to accept and parse dictionary of kwargs
* add functionality to automated analyses as they are completed

### PR DESCRIPTION
**Pull request by cadeduckworth**
close #237

### CODE REVIEW

**Comment by cadeduckworth** [comment]
This PR will require that PR#217 or PR#227 is merged to complete testing of directory iteration function and workflows registry module.

**Comment by cadeduckworth** [comment]
Requires merging of recent updates from `develop` before continuing work on tests.

**Comment by orbeckst** [comment]
There are test failures related to SM25/SM26. Let me know when the PR is ready for review (or if you need input on something specific).

**Comment by cadeduckworth** [comment]
@orbeckst I added the workflows testing data path to the STATES dictionary in __init__.py and cleaned up the testing module by making the necessary changes and adding a few more fixtures. Once it is confirmed that everything looks good I will go through the docs once more to revise/finalize!

**Comment by cadeduckworth** [comment]
UPDATE: I forgot to change the single quotes to double quotes to remove the backslash from logger/error messages in `base.py` and `test_worflows_base.py`. 

Also, the `dir_paths.csv` file in the testing resources still might need a change of paths to a relative convention, since it assumes that the CI is running the tests from outside of the `mdpow` directory using an absolute path, and paths can always change, which would break the test.

**Review by orbeckst — CHANGES_REQUESTED** [review]
Looking really good, I don't see any problems.

EDIT: Please also add some documentation for the registry.

Just add an Enhancement entry to CHANGES. (Please also fix the year of the 0.8.0 release to 2022 instead of 2021).

**Comment by cadeduckworth** [comment]
@orbeckst 

requested changes addressed

File: mdpow/workflows/registry.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+=======================================================================================
+
+Each entry in :mod:`mdpow.workflows.registry` corresponds to an :class:`~mdpow.analysis.ensemble.EnsembleAnalysis` for which exists a corresponding
+automated workflow.
+
+Intended for use with :mod:`mdpow.workflows.base` to specify which
+:class:`~mdpow.analysis.ensemble.EnsembleAnalysis` should be iterated over
+the provided project data.
+
+.. seealso:: :mod:`~mdpow.workflows.base`
```
Comment (orbeckst) [inline mdpow/workflows/registry.py:15]: make this a single seealso (it can do lists, check syntax)

File: mdpow/workflows/registry.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
+# MDPOW: registry.py
+# 2023 Cade Duckworth
+
+"""
+:mod:`mdpow.workflows.registry` --- Registry of currently supported automated workflows
+=======================================================================================
+
+Each entry in :mod:`mdpow.workflows.registry` corresponds to an :class:`~mdpow.analysis.ensemble.EnsembleAnalysis` for which exists a corresponding
+automated workflow.
```
Comment (orbeckst) [inline mdpow/workflows/registry.py:9]: Here should be a list of workflows, essentially manually maintain what's in registry.

File: mdpow/workflows/registry.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+"""
+
+# import analysis
+from mdpow.workflows import dihedrals
+ 
+registry = {
+    'DihedralAnalysis' : dihedrals.automated_dihedral_analysis,
+    'SolvationAnalysis' : None,
+    'HBondAnalysis' : None
```
Comment (orbeckst) [inline mdpow/workflows/registry.py:26]: remove anything that's not present yet

File: mdpow/workflows/registry.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+.. seealso:: :mod:`~mdpow.workflows.base`
+.. seealso:: :mod:`~mdpow.workflows.dihedrals`
+
+"""
+
+# import analysis
+from mdpow.workflows import dihedrals
+ 
+registry = {
```
[the remark this answers is not in the document]
Comment (orbeckst) [reply mdpow/workflows/registry.py:23]: State here for developers what needs to be added, i.e., function/class

File: mdpow/workflows/base.py [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+       
+       Typical Workflow::
+
+           directory_paths = directory_paths(parent_directory='/foo/bar/MDPOW_projects')
+           directory_iteration(directory_paths)
+
+           or
+
+           directory_paths = directory_paths(csv='/foo/bar/MDPOW.csv')
+           directory_iteration(directory_paths)
```
Comment (orbeckst) [inline mdpow/workflows/base.py:63]: do proper code formatting

File: mdpow/workflows/base.py [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+            'path': locations
+        }
+    )
+        if csv_save_dir is not None:
+            
+            directory_paths.to_csv(f'{csv_save_dir}/dir_paths.csv', index=False)
+            logger.info(f'dir_paths saved under {csv_save_dir}')
+
+        else:
+            current_directory = os.system('pwd')
```
Comment (orbeckst) [inline mdpow/workflows/base.py:96]: Don't use `os.system` — use a python function such as `cwd` or something from `pathlib`

File: mdpow/workflows/base.py [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+        locations = pd.read_csv(csv)
+        directory_paths = locations.sort_values(by=['molecule', 'resname', 'path']).reset_index(drop=True)
+
+    return directory_paths
+
+def directory_iteration(directory_paths, ensemble_analysis, **kwargs):
+    """Takes a :class:`pandas.DataFrame` created by
+       :func:`~mdpow.workflows.base.directory_paths`
+       and iterates over the project paths to implement
+       :func:`~mdpow.workflows.dihedrals.automated_dihedral_analysis`
```
Comment (orbeckst) [inline mdpow/workflows/base.py:110]: write docs more generally, not just dihedrals

File: mdpow/workflows/base.py [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+
+       *directory_paths*
+           :class:`pandas.DataFrame` that provides paths to MDPOW projects
+
+       *ensemble_analysis*
+           name of the :class:`~mdpow.analysis.ensemble.EnsembleAnalysis`
+           that corresponds to the desired automated workflow module
+
+       *kwargs*
+           keyword arguments from :func:`~mdpow.workflows.dihedrals.automated_dihedral_analysis`
```
Comment (orbeckst) [inline mdpow/workflows/base.py:125]: write docs more generally, not just dihedrals

File: mdpow/workflows/base.py [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+    elif csv is not None:
+        locations = pd.read_csv(csv)
+        directory_paths = locations.sort_values(by=['molecule', 'resname', 'path']).reset_index(drop=True)
+
+    return directory_paths
+
+def directory_iteration(directory_paths, ensemble_analysis, **kwargs):
+    """Takes a :class:`pandas.DataFrame` created by
+       :func:`~mdpow.workflows.base.directory_paths`
+       and iterates over the project paths to implement
```
Comment (orbeckst) [inline mdpow/workflows/base.py:109]: not "implement" but "run"

File: mdpow/workflows/base.py [hunk 10]
```diff
@@ -0,0 +1,166 @@
+# MDPOW: base.py
+# 2022 Cade Duckworth
+
+"""
+:mod:`mdpow.workflows.base` --- Automated workflow base functions
+=================================================================
+
```
Comment (orbeckst) [inline mdpow/workflows/base.py:7]: Add text to explain (primarily for developers) what you need to do to make this module work. Explain (and link) to registry.

File: mdpow/workflows/base.py [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+            directory_paths.to_csv('dir_paths.csv', index=False)
+            logger.info(f'dir_paths saved under {current_directory}')
+
+    elif csv is not None:
+        locations = pd.read_csv(csv)
+        directory_paths = locations.sort_values(by=['molecule', 'resname', 'path']).reset_index(drop=True)
+
+    return directory_paths
+
+def directory_iteration(directory_paths, ensemble_analysis, **kwargs):
```
Comment (orbeckst) [inline mdpow/workflows/base.py:106]: The name `directory_iteration` is not telling me that we are running an EnsembleAnalysis for multiple MDPOW runs. Come up with a better name.

File: mdpow/workflows/base.py [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+
+    except TypeError as err:
+        msg = (f"Invalid ensemble_analysis {ensemble_analysis}. An EnsembleAnalysis type that "
+                "corresponds to an existing automated workflow module must be input as a kwarg. "
+                "ex: ensemble_analysis='DihedralAnalysis'")
+        logger.error(f'workflow module for {ensemble_analysis} does not exist yet')
+
+        raise TypeError(msg)
+
+    return logger.info('all analyses completed')
```
Comment (orbeckst) [inline mdpow/workflows/base.py:166]: Returning logger does not make sense.

I think we should be returning a collection of the data that's produced — eg all filenames. How will you know after this has run, what you got? It should _not_ require knowing a file naming scheme.

**Review by orbeckst — CHANGES_REQUESTED** [review]
I found a bunch of doc things and small code stuff.

EDIT (after our discussion): Please address the "small stuff" in the comments. For the bigger issue below (**maintaining a database of all produced artifacts/results**), please _open a new issue_:

The bigger conceptual issue is that with the current automated workflows we dump a whole bunch of files into the file system without recording them. I would like your automated analysis to return a manifest of files so that downstream code can work with these data structures to do other things — eg compile them into PDFs, load CSVs for further processing etc. It shouldn't be necessary to reconstruct file names by rules that are embodied inside your code.

This means that any automated workflow should be ultimately returning something like a dict or list (or df) of "type_of_output -> filename". 

The `directory_iteration()` should then bundle everything together, probably in a big dict or better a df that allows one to "get all dihedral KDE plots that contain octanol" — I imagine a df with many columns (molecule, solvent, quantity, datatype,... , filename) where datatype could be "image" or "data" (for CSV) or similar. As part of `base` you'd have to _define_ this API.

File: mdpow/workflows/base.py (hunk 12 above)
Comment (orbeckst) [reply mdpow/workflows/base.py:166]: For right now, please change to 
```python
logger.info('all analyses completed')
return
```
(and elsewhere, too)

File: mdpow/workflows/base.py (hunk 5 above)
Comment (cadeduckworth) [reply mdpow/workflows/base.py:63]: Also, can you look at the `.. autodata:: registry` portion of registry.py and let me know if there is a way to suppress the type hint. If you look at the docs build it included instructions on how to build a dictionary underneath the dictionary values for registry.

**Comment by cadeduckworth** [comment]
@orbeckst 

I think I properly addressed all of the requested changes. I will come back and review the documents with fresh eyes later. Also, see question above.

**Comment by orbeckst** [comment]
@cadeduckworth can you please update the branch with the main branch and please raise the issue about generating manifest dataframe for all generated data/images.

File: mdpow/workflows/base.py [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+
+To analyze multiple MDPOW projects, provide :func:`project_paths`
+with the top-level directory containing all MDPOW projects' simulation data
+to obtain a :class:`pandas.DataFrame` containing the project information
+and paths. Then, :func:`automated_project_analysis` takes as input the
+aforementioned :class:`pandas.DataFrame` and runs the specified
+:class:`~mdpow.analysis.ensemble.EnsembleAnalysis` for all MDPOW projects
+under the top-level directory provided to :func:`project_paths`.
+
+Currently supported workflows - :mod:`~mdpow.workflows.dihedrals`
```
Comment (orbeckst) [inline mdpow/workflows/base.py:16]: not needed here
```suggestion
```

File: mdpow/workflows/base.py [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+            res_temp = loc.strip().split('/')
+            resnames.append(res_temp[-1])
+
+        project_paths = pd.DataFrame(
+        {
+            'molecule': resnames,
+            'resname': resnames,
+            'path': locations
+        }
+    )
```
Comment (orbeckst) [inline mdpow/workflows/base.py:101]: indent

File: mdpow/workflows/base.py [hunk 15]
```diff
@@ -0,0 +1,180 @@
...
+        if csv_save_dir is not None:
+
```
Comment (orbeckst) [inline mdpow/workflows/base.py:103]: can remove some blank lines
```suggestion
```

File: mdpow/workflows/base.py [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+            'molecule': resnames,
+            'resname': resnames,
+            'path': locations
+        }
+    )
+        if csv_save_dir is not None:
+            
+            project_paths.to_csv(f'{csv_save_dir}/project_paths.csv', index=False)
+            logger.info(f'project_paths saved under {csv_save_dir}')
+
```
Comment (orbeckst) [inline mdpow/workflows/base.py:106]: make a bit more compact
```suggestion
```

File: mdpow/workflows/base.py [hunk 17]
```diff
@@ -0,0 +1,180 @@
...
+       Typical Workflow::
+
+           project_paths = project_paths(parent_directory='/foo/bar/MDPOW_projects')
+           automated_project_analysis(project_paths)
+
+           'or'
+
+           project_paths = project_paths(csv='/foo/bar/MDPOW.csv')
+           automated_project_analysis(project_paths)
+
```
Comment (orbeckst) [inline mdpow/workflows/base.py:76]: If you look at https://mdpow--229.org.readthedocs.build/en/229/workflows/base.html#mdpow.workflows.base.project_paths
image
you see that the "or" is still formatted as code. 

The formatting below should fix it:

```suggestion
       Typical Workflow::
           project_paths = project_paths(parent_directory='/foo/bar/MDPOW_projects')
           automated_project_analysis(project_paths)
           
        or::
        
           project_paths = project_paths(csv='/foo/bar/MDPOW.csv')
           automated_project_analysis(project_paths)
           
```

File: mdpow/workflows/base.py [hunk 18]
```diff
@@ -0,0 +1,180 @@
...
+       *ensemble_analysis*
+           name of the :class:`~mdpow.analysis.ensemble.EnsembleAnalysis`
+           that corresponds to the desired automated workflow module
+
+       *kwargs*
+           keyword arguments for the supported automated workflows,
+           currently - :func:`~mdpow.workflows.dihedrals.automated_dihedral_analysis`
+
+           .. autodata:: mdpow.workflows.dihedrals.automated_dihedral_analysis
```
Comment (orbeckst) [inline mdpow/workflows/base.py:139]: remove the details on automated_dihedral_analysis
```suggestion
           keyword arguments for the supported automated workflows, see the
           :mod:`~mdpow.workflows.registry` for all available workflows and their
           call signatures
```
(also, the `autodata` messes up formatting — I am not sure what you actually wanted to show here, if you want to link, just do a 
```rest
:func:`mdpow.workflows.dihedrals.automated_dihedral_analysis`
```

File: mdpow/workflows/base.py [hunk 19]
```diff
@@ -0,0 +1,180 @@
...
+
+    try:
+        for row in project_paths.itertuples():
+            molname = row.molecule
+            resname = row.resname
+            dirname = row.path
+
+            logger.info(f'starting {molname}')
+
+            registry.registry[ensemble_analysis](dirname=dirname, resname=resname, molname=molname, **kwargs)
```
Comment (orbeckst) [inline mdpow/workflows/base.py:159]: Wrap _only_ this line in try/except.

The smaller the code that you're bracketing, the more precisely you are still able to locate other errors.

File: mdpow/workflows/registry.py [hunk 20]
```diff
@@ -0,0 +1,31 @@
...
+# 2023 Cade Duckworth
+
+"""
+:mod:`mdpow.workflows.registry` --- Registry of currently supported automated workflows
+=======================================================================================
+
+Each entry in :mod:`mdpow.workflows.registry` corresponds to an :class:`~mdpow.analysis.ensemble.EnsembleAnalysis`
+for which exists a corresponding automated workflow.
+
+Currently supported workflows - :mod:`~mdpow.workflows.dihedrals`
```
Comment (orbeckst) [inline mdpow/workflows/registry.py:11]: Does not format as a list, see https://mdpow--229.org.readthedocs.build/en/229/workflows/registry.html#mdpow-workflows-registry-registry-of-currently-supported-automated-workflows
image

Can you please create a [reST table](https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#tables) (and see more under [table role](https://docutils.sourceforge.io/docs/ref/rst/directives.html#table)) that looks like

| **keyword** | **analysis function** |
|---------|---------|
| DihedralAnalysis | :func:'dihedrals.automated_dihedral_analysis <mdpow.workflows.dihedrals.automated_dihedral_analysis>' |

where the keyword is the string needed for :func:'mdpow.workflows.base.automated_project_analysis' and the analysis function is the actual automated function that will be executed — essentially the same information that is in the registry dict but immediately understandable (and usable) by humans.

For the reST, should look something like (include [name](https://docutils.sourceforge.io/docs/ref/rst/directives.html#common-options) for hyperlinking)
```reST
.. table:: Available automated workflows.
   :widths: auto
   :name: workflows_table

    ==============  ============
    keyword         function
    ==============  ============
    ...
    ...
    ==============  ============
```

File: mdpow/workflows/registry.py [hunk 21]
```diff
@@ -0,0 +1,31 @@
...
+
+Intended for use with :mod:`mdpow.workflows.base` to specify which
+:class:`~mdpow.analysis.ensemble.EnsembleAnalysis` should run iteratively over
+the provided project data directory. To include a new automated workflow for use with
+:mod:`mdpow.workflows.base`, create a key that is the name of the corresponding
+:class:`~mdpow.analysis.ensemble.EnsembleAnalysis`, with the value defined as
+`<workflow module name>.<top-level function>`.
+
+.. autodata:: registry
```
Comment (orbeckst) [inline mdpow/workflows/registry.py:20]: you can probably use `:nomember:` or other options — check the docs https://www.sphinx-doc.org/en/master/usage/restructuredtext/index.html

File: mdpow/workflows/registry.py [hunk 22]
```diff
@@ -0,0 +1,31 @@
...
+
+.. seealso:: :mod:`~mdpow.workflows.base`, :mod:`~mdpow.workflows.dihedrals`
```
Comment (orbeckst) [inline mdpow/workflows/registry.py:22]: only include base

File: mdpow/workflows/registry.py [hunk 23]
```diff
@@ -0,0 +1,31 @@
...
+
+"""
+
+# import analysis
+from mdpow.workflows import dihedrals
+ 
+registry = {
```
Comment (orbeckst) [inline mdpow/workflows/registry.py:29]: Add documentation for the `registry`. Either as part of the code
```python
#: The `registry` contains ... 
#: New workflows are added by ...
#:    The currently available workflows are listed in  the :ref:`Table of Workflows <workflows_table>`
registry = { ... }
```
or by adding the text above when writing 
```rest
.. :data:`registry`
   The `registry` contains ...   New workflows are added by ...
   The currently available workflows are listed in  the :ref:`Table of Workflows <workflows_table>`
```

(note the hyperlink ref to the table's name)

**Review by orbeckst — CHANGES_REQUESTED** [review]
Please see comments for requested improvements. Thanks!

**Comment by orbeckst** [comment]
@cadeduckworth please ping me when you're ready for review

File: mdpow/workflows/registry.py (hunk 23 above)
Comment (cadeduckworth) [reply mdpow/workflows/registry.py:45]: @orbeckst 

I can't figure this part out. I am having trouble writing the `..  data::` or `:data:` section in a way that is recognized by `.. autodata::`.

The docs are very confusing when it comes to documenting objects within a module.

**Comment by cadeduckworth** [comment]
@orbeckst 
ready for review

**Review by orbeckst — APPROVED** [review]
looks good, let me just merge to doc changes.

**Comment by orbeckst** [comment]
Hooray 🎉 , it's merged @cadeduckworth !

**Pull request merged**